content/en/docs/contributing/coding-conventions.md

```diff
@@ -0,0 +1,823 @@
+1. Proposal MP: Guideline when using UK_ASSERT and when using UK_CRASH
+	UK_CRASH is for situations where you can not formalize an assertion
+  	Something like UK_ASSERT(false)
+1. MS: Some should be present in release builds
+UK_BUG_ON should never be removed from the code (with condition)
+The same for UK_CRASH (without condition)
+Change behavior of UK_BUG_ON to always be part of the code, even in release builds
+
+### Parameter Validation
+
+Do rigorous parameter validation.
+However, parameter validation should be done using assertions where the parameter can be expected to be not directly determined by external input.
+This means in particular that `uk_`, `ukplat_`, and `ukarch_` APIs should use `UK_ASSERT()`:
+
+```c
+int uk_cc_myfunc(int bobo)
+{
+    UK_ASSERT(bobo == 0xdead);
+
+    /* do something */
+}
+✅
+```
+
+```c
+int uk_cc_myfunc(int bobo)
+{
+    if (unlikely(bobo == 0xdead))
+            return -EINVAL;
+
+    /* do something */
+}
+❌
+```
+
+This is in contrast to, for example, POSIX, where explicit parameter validation is expected.
+In a Unikernel, though, we see the OS components as an (internal) extension to the application.
+Users of the APIs are thus expected to make sure only sane parameters are used just like with internal application functions.
+
+Generally, use the `unlikely()` macro when checking for error cases (only). 
+It is defined in `<uk/essentials.h>`.
+
+When working with the low-level Unikraft API, it is essential to validate parameters using UK_ASSERT when specific parameter ranges need to be enforced. In cases where there is no explicit specification, performing an error check is recommended to ensure parameter validity.
+
+When working with higher-level APIs, parameter validation should be handled by returning appropriate error codes according to the relevant standard (such as POSIX).
+
+### Comment Keywords
+
+1. `NOTE`: Description of how the code works (when it isn't self evident)
+1. `XXX:` Warning about possible pitfalls, can be used as `NOTE:XXX:`
+1. `HACK`: Not very well written or malformed code to circumvent a problem/bug
+Should be used as HACK:FIXME:
+1. `FIXME`: This works, sort of, but it could be done better. (usually code written in a hurry that needs rewriting)
+1. `BUG`: There is a problem here
+1. `TODO`: No problem, but additional code needs to be written, usually when you are skipping something
+1. `UNDONE`: A reversal or "roll back" of previous code
+1. `OPTIMIZE`: Code can / should be optimized
+
+### Magic Numbers
+
+These are some magic numbers used in the Unikraft code:
+
+1. `UKPLAT_BOOTINFO_MAGIC` 0xB007B0B0 (Boot Bobo)
+1. `FDT_MAGIC 0xd00dfeed` (FDT specification)
+1. `UK_TP_HEADER_MAGIC` 0x64685254 (TRhd)
+1. `UK_TP_DEF_MAGIC` 0x65645054 (TPde)
+1. `CL_UKTLS_SANITY_MAGIC` 0xb0b0f00d (Bobo food)
+1. `0xDEADB0B0`, `0xB0B0CAFE ` (in ukvmem)
+1. `MULTIBOOT_HEADER_MAGIC` 0x1BADB002
+1. `MULTIBOOT_BOOTLOADER_MAGIC` 0x2BADB002
+```
+
```

Comment:
Also add that when used in source code, the symbol of the magic number should be used, never the value.